Fix timeout support #41

crankydillo opened the pull request:
See the issue for more info, but basically the code I'm replacing
assumed that:

1. Connect/read timeout was always done with client creation.
2. Our impl (technically feign) couldn't successfully set those per
   request.

This work addresses the above and 3 issues:

1. How a client can specify how the timeouts are specified.
2. How the builder chain can honor the above in a nice way (i.e if you
   a client says it doesn't support request timeouts then those aren't
   exposed via the build chain.
3. How to test without repetition.

So, it was just interesting to try and solve this.  Now, I don't think
this code is super-complicated, but I can understand distaste for this.
I primarily did it just out of interest:)  I think you could make a
decent argument that we just bring back the timeout methods to all
builder chains and in the jaxrs2+client (we could still add a builder
option), we'd just log that these are ignored.  That's unfortunate, but
if this is the only case where we need to ignore request-level timeouts,
then I can understand not wanting this code.

Commit:
Fix timeout support #41

See the issue for more info, but basically the code I'm replacing
assumed that:

1. Connect/read timeout was always done with client creation.
2. Our impl (technically feign) couldn't successfully set those per
   request.

This work addresses the above and 3 issues:

1. How a client can specify how the timeouts are specified.
2. How the builder chain can honor the above in a nice way (i.e if you
   a client says it doesn't support request timeouts then those aren't
   exposed via the build chain.
3. How to test without repetition.

So, it was just interesting to try and solve this.  Now, I don't think
this code is super-complicated, but I can understand distaste for this.
I primarily did it just out of interest:)  I think you could make a
decent argument that we just bring back the timeout methods to all
builder chains and in the jaxrs2+client (we could still add a builder
option), we'd just log that these are ignored.  That's unfortunate, but
if this is the only case where we need to ignore request-level timeouts,
then I can understand not wanting this code.

crankydillo commented:
@fabiocarvalho777 I pushed some more javadoc.  As we discussed, the main user impact here is to the builder.  At a certain point, you supply a client (or use `defaultClient()`) where the client implementations leverage marker interfaces to determine whether or not the builder should expose timeouts or not.  Most do expose timeouts.  Currently, only JAX-RS does not.

This is not feature complete.  I need to do something similar for async, finish up testing, and adjust user doc.  However, the main idea is there, and I thought this might be controversial, so I didn't want to do any needless work.  This was supposed to be opened as a draft, but I guess didn't do that the 2nd go around.

fabiocarvalho777 commented on mocca-jaxrs2/src/main/java/com/paypal/mocca/client/MoccaJaxrsClient.java:
```diff
@@ -36,8 +36,8 @@
-        log.debug("Users of this may attempt to set read timeout and connect timeout per request. " +
-            "However, those are ignored. They should be established in the supplied javax.ws.rs.Client.");
+        log.info("Mocca implementation _may_ attempt to set read timeout and connect timeout per request.  " +
+            "However, those are ignored.  They should be established in the supplied javax.ws.rs.Client.");
```
@crankydillo out of curiosity, why do you always add two spaces after a period?

crankydillo replied:
Standard (American) English rules:)

fabiocarvalho777 commented on mocca-jaxrs2/src/main/java/com/paypal/mocca/client/MoccaJaxrsClient.java:
```diff
@@ -36,8 +36,8 @@
         super(new JAXRSClient(new StubbornClientBuilder(client)));
-        log.debug("Users of this may attempt to set read timeout and connect timeout per request. " +
-            "However, those are ignored. They should be established in the supplied javax.ws.rs.Client.");
+        log.info("Mocca implementation _may_ attempt to set read timeout and connect timeout per request.  " +
```
I would prefer to not have this log message, since there is no attempt to change timeouts right here. I believe a message like this might confuse sers, it would be better to just log a message like this when an attempt to change a timeout actually happens.

crankydillo replied:
That means you will always get that message for every request.  Of course, people can disable that, but they'll need to go do that explicitly.  I guess I could set it to trace.  Anyhow, let me just remove this altogether now.  It's not crystal clear this is needed.

fabiocarvalho777 commented on mocca-jaxrs2/build.gradle:
```diff
@@ -7,5 +7,6 @@ dependencies {
                        lib.jersey_client,
-                       lib.jersey_hk2
+                       lib.jersey_hk2,
+                       lib.activation
```
Which line of code, or situation, required the inclusion of this dependency?

crankydillo replied:
Honestly, I don't recall.  I thought the build basically told me to, but I just removed it and am still getting green builds with `./gradlew clean test`.  I will remove `lib.activation`.

fabiocarvalho777 commented on mocca-client/src/main/java/com/paypal/mocca/client/MoccaClient.java:
```diff
@@ -208,47 +213,116 @@ private SyncBuilder(final String serverBaseUrl) {
+             *
+             * @see WithoutRequestTimeouts
+             */
+            public class WithRequestTimeouts extends Base {
```
Should this be package private, like `WithoutRequestTimeouts`?

fabiocarvalho777 replied:
I am confused with the name of this class. As I review this PR, I have always to come back here to try to understand what it means. Let me see if I can explain what I mean:\

Correct me if I am wrong but, regardless of this type, clients will always be "with request timeouts". The difference is whether they are supposed to be configured using Mocca API or the client native API.

If I understood this right, clients of this type will only let users set timeouts using Mocca API. Is that right?

crankydillo replied:
Fabio and I spoke about this, but for posterity...

`WithRequestTimeouts` and `WithoutRequestTimouts` are just builder paths.  The former exposes Mocca builder-based timeout establishment.  I'm not a huge fan of the names either, but it's not really anticipated that clients 'see' these types very much.  I assume they'll just chain the builder methods.

Clients will not always send you down the `WithRequestTimeouts` path.  The client passed in will either support Mocca builder-based timeouts or not (most currently do).  The ones that do implement an interface that sends you down the `WithRequestTimeouts` path where they can set the socket and connect timeout with Mocca builder methods.  The ones that don't (jaxrs) send you down the `WithoutRequestTimeouts` path where they can't.  For the latter, we assume that native client functionality is used to set the timeouts.  The reason for this trickery is that our underlying client facade will try to set default timeout values so any `MoccaHttpClient` we feed in must be able to ignore that.

fabiocarvalho777 commented on mocca-client/src/main/java/com/paypal/mocca/client/MoccaClient.java:
```diff
@@ -208,47 +213,116 @@ private SyncBuilder(final String serverBaseUrl) {
-                builder = builder.contract(new MoccaFeignContract())
-                    .encoder(encoder)
-                    .decoder(new MoccaFeignDecoder());
+                WithRequestTimeouts options(
```
I am confused. Is this method supposed to be exposed to end-users via Mocca API? If so, shouldn't it be public?

fabiocarvalho777 commented on mocca-client/src/main/java/com/paypal/mocca/client/MoccaClient.java:
```diff
@@ -208,47 +213,116 @@ private SyncBuilder(final String serverBaseUrl) {
-                    .encoder(encoder)
-                    .decoder(new MoccaFeignDecoder());
+                WithRequestTimeouts options(
+                    final Duration connectTimeout,
```
What if end user only wants to provide `connectTimeout`?

crankydillo replied:
They are out of luck?  :)

Commit:
The main part of this commit is covering tests for cases where...

Mocca will govern timeout via its API.  Also includes some review
comments like access modifiers, javadoc, and gradle snafus.

Still have async test cases to write.

Commit:
Address timeout issue for async cases #41

Includes generics fixes to the optional context for feign's async
client.  Honestly, we probably need to hide that in some way because I
don't think we expose it..

I addressed one PMD issue that I understand, but I didn't really get the
transient thing.  Those aren't beans..

Commit:
Test fixes

I thought these got added with rebase:(

crankydillo commented:
@fabiocarvalho777 This is my proposed fix.  It does include a few other fixes, primarily related to the `C` in feign's `AsyncClient<C>`.

I did not update the end user doc.  I am hoping you will be willing to do that after we decide if this is what we want to do.

I will fix any reasonable CI failures.

Commit:
Remove unused async context type variable #41
